vms/evm/uptimetracker/state.go

```diff
@@ -0,0 +1,234 @@
+		}
+
+		s.addValidator(validationID, vdr)
+	}
+
+	if err := it.Error(); err != nil {
+		return nil, fmt.Errorf("failed to iterate: %w", err)
+	}
+
+	return s, nil
+}
+
+func (s *state) GetUptime(nodeID ids.NodeID) (time.Duration, time.Time, error) {
+	v, ok := s.getValidatorByNodeID(nodeID)
+	if !ok {
+		return 0, time.Time{}, database.ErrNotFound
+	}
+
+	return v.UpDuration, time.Unix(int64(v.LastUpdated), 0), nil
+}
+
+func (s *state) SetUptime(
+	nodeID ids.NodeID,
+	upDuration time.Duration,
+	lastUpdated time.Time,
+) error {
+	v, ok := s.getValidatorByNodeID(nodeID)
+	if !ok {
+		return database.ErrNotFound
+	}
+
+	v.UpDuration = upDuration
+	v.LastUpdated = uint64(lastUpdated.Unix())
+
+	s.updatedValidators.Add(v.validationID)
+
+	return nil
+}
+
+func (s *state) GetStartTime(nodeID ids.NodeID) (time.Time, error) {
+	v, ok := s.getValidatorByNodeID(nodeID)
+	if !ok {
+		return time.Time{}, database.ErrNotFound
+	}
+
+	return time.Unix(int64(v.LastUpdated), 0), nil
+}
+
+func (s *state) addValidatorUpdate(vdr *validator) {
+	s.addValidator(vdr.validationID, vdr)
+	s.updatedValidators.Add(vdr.validationID)
+}
+
+func (s *state) updateValidator(validationID ids.ID, isActive bool) bool {
+	v := s.validators[validationID]
+
+	if v.IsActive == isActive {
+		return false
+	}
+
+	v.IsActive = isActive
+	s.updatedValidators.Add(validationID)
+	return true
+}
+
+func (s *state) deleteValidator(validationID ids.ID) bool {
+	v, ok := s.validators[validationID]
+	if !ok {
+		return false
+	}
+
+	delete(s.validators, v.validationID)
+	delete(s.nodeIDsToValidationIDs, v.NodeID)
+
+	s.deletedValidators.Add(v.validationID)
+
+	return true
+}
+
+func (s *state) writeState() error {
+	batch := s.db.NewBatch()
+
+	for validationID := range s.updatedValidators {
+		validatorBytes, err := codecManager.Marshal(
+			codecVersion,
+			s.validators[validationID],
+		)
+		if err != nil {
+			return fmt.Errorf("failed to marshal validator: %w", err)
+		}
+
+		if err := batch.Put(validationID[:], validatorBytes); err != nil {
+			return fmt.Errorf("failed to put validator: %w", err)
+		}
+	}
+
+	for validationID := range s.deletedValidators {
+		if err := batch.Delete(validationID[:]); err != nil {
+			return fmt.Errorf("failed to delete validator: %w", err)
+		}
+	}
+
+	if err := batch.Write(); err != nil {
+		return fmt.Errorf("failed to write batch: %w", err)
+	}
+
+	// We have written all pending updates
+	clear(s.updatedValidators)
+	clear(s.deletedValidators)
+
+	return nil
+}
+
+func (s *state) addValidator(validationID ids.ID, validator *validator) {
+	s.validators[validationID] = validator
+	s.nodeIDsToValidationIDs[validator.NodeID] = validationID
+}
+
+func (s *state) getValidatorByNodeID(nodeID ids.NodeID) (*validator, bool) {
+	validationID, ok := s.nodeIDsToValidationIDs[nodeID]
+	if !ok {
+		return nil, false
+	}
+
+	// s.nodeIDsToValidationIDs is guaranteed to have the same data as in s.validators
```

Comment:
Comments have a maintenance cost associated with them so we should avoid coupling code identifiers to comments - this stuff isn't indexed by the IDE and even if it was it's hard to keep in-sync w/ the code. 
```suggestion
	// we are guaranteed to have this validator
```